### PR TITLE
Add fertility to count coverage

### PR DESCRIPTION
**Pull request by bricksdont**
This PR adds a missing piece from this paper:

https://arxiv.org/abs/1601.04811

namely, fertility for count coverage.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/coverage.py [hunk 1]
```diff
@@ -33,15 +33,18 @@ class CoverageConfig(config.Config):
     Coverage configuration.
 
     :param type: Coverage name.
+    :param max_fertility: Maximum number of target words generated by a source word.
     :param num_hidden: Number of hidden units for coverage networks.
     :param layer_normalization: Apply layer normalization to coverage networks.
     """
     def __init__(self,
                  type: str,
+                 max_fertility: int,
```
Comment (fhieber) [inline sockeye/coverage.py]: Could you make this a keyword argument with a default value (2) to ensure compatibility with older models using coverage?

File: sockeye/coverage.py [hunk 2]
```diff
@@ -53,14 +56,16 @@ def get_coverage(config: CoverageConfig) -> 'Coverage':
     :param config: Coverage configuration.
     :return: Instance of Coverage.
     """
-    if config.type == 'count':
-        utils.check_condition(config.num_hidden == 1, "Count coverage requires coverage_num_hidden==1")
+    if config.type == 'count' or config.type == 'fertility':
```
Comment (fhieber) [inline sockeye/coverage.py]: if you are at it, it'd be great if you could change to string comparisons to use the newly introduced constants defined in constants.py

File: sockeye/coverage.py [hunk 3]
```diff
@@ -108,7 +113,6 @@ def __init__(self) -> None:
     def on(self, source: mx.sym.Symbol, source_length: mx.sym.Symbol, source_seq_len: int) -> Callable:
         """
         Returns callable to be used for updating coverage vectors in a sequence decoder.
-
```
Comment (fhieber) [inline sockeye/coverage.py]: [super nit] delete the deleted newline :)

File: sockeye/coverage.py [hunk 4]
```diff
@@ -129,6 +133,67 @@ def update_coverage(prev_hidden: mx.sym.Symbol,
...
+
+        :param source: Shape: (batch_size, seq_len, encoder_num_hidden).
+        :param source_length: Shape: (batch_size,).
+        :param source_seq_len: Maximum length of source sequences.
+        :return: Coverage callable.
+        """
+
+        # (batch_size, seq_len, 1)
+        source_fertility = mx.sym.FullyConnected(data=source,
+                                              weight=self.cov_e2f_weight,
```
Comment (fhieber) [inline sockeye/coverage.py]: indentation

File: sockeye/coverage.py [hunk 5]
```diff
@@ -129,6 +133,67 @@ def update_coverage(prev_hidden: mx.sym.Symbol,
...
+            """
+            :param prev_hidden: Previous hidden decoder state. Shape: (batch_size, decoder_num_hidden).
+            :param attention_prob_scores: Current attention scores. Shape: (batch_size, source_seq_len).
+            :param prev_coverage: Shape: (batch_size, source_seq_len, coverage_num_hidden).
+            :return: Updated coverage matrix . Shape: (batch_size, source_seq_len, coverage_num_hidden).
+            """
+
+            # (batch_size, seq_len, 1)
+            fertility = mx.sym.Activation(data=source_fertility,
+                                         act_type="sigmoid",
```
Comment (fhieber) [inline sockeye/coverage.py]: indendation

File: sockeye/coverage.py [hunk 6]
```diff
@@ -129,6 +133,67 @@ def update_coverage(prev_hidden: mx.sym.Symbol,
...
+                                         name="%sactivation" % self.prefix)
+
+            # (batch_size, seq_len, 1)
+            scaled_fertility = 1 / (self.max_fertility * fertility)
```
Comment (fhieber) [inline sockeye/coverage.py]: could the activation and the scaling be moved outside `udpate_coverage()` and directly after the FullyConnected? They only have to be defined once per batch right?

Comment (bricksdont) [reply sockeye/coverage.py]: Good point - yes, if I understand correctly they can be computed once.

**Pull request merged**